File: simservice/PandaPowerWrapper/simulation_event_emitter.py
import json
import fcntl
import logging
import os
import threading
import uuid
from datetime import datetime, timezone
from pathlib import Path


class SimulationEventEmitter:
    EVENT_TOPIC = "simservice.logs.events"
    _file_lock = threading.RLock()

    def __init__(self, task_id, source, component, producer=None, results_dir=None):
        self.task_id = task_id
        self.source = source
        self.component = component
        self.producer = producer
        self.results_dir = Path(results_dir) if results_dir else None

    def emit(self, category, event_code, message, progress=None, simulation_time=None, metadata=None):
        if not self.task_id:
            return

        event = {
            "id": f"evt_{uuid.uuid4()}",
            "schema_version": "1.0",
            "task_id": self.task_id,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "source": self.source,
            "category": category,
            "component": self.component,
            "event_code": event_code,
            "message": message,
        }
        if progress is not None:
            event["progress"] = progress
        if simulation_time is not None:
            event["simulation_time"] = simulation_time
        if metadata is not None:
            event["metadata"] = metadata

        self._persist_event(event, category == "METRIC")
        self._publish(event)

    def _persist_event(self, event, metric):
        if not self.results_dir:
            return
        try:
            directory = self.results_dir / "events"
            directory.mkdir(parents=True, exist_ok=True)
            with self._file_lock, (directory / ".writer.lock").open("a+b") as lock:
                os.chmod(directory / ".writer.lock", 0o666)
                fcntl.lockf(lock, fcntl.LOCK_EX)
                try:
                    self._append_jsonl("structured_events.jsonl", event)
                    if metric:
                        self._append_jsonl("metrics.jsonl", event)
                finally:
                    fcntl.lockf(lock, fcntl.LOCK_UN)
        except OSError:
            logging.getLogger(__name__).warning("Local event persistence failed; attempting Kafka", exc_info=True)

    def metric_snapshot(self, simulation_time, metrics, progress=None, metadata=None):
        meta = {"metrics": metrics}
        if metadata:
            meta.update(metadata)
        self.emit(
            "METRIC",
            "METRIC_SNAPSHOT",
            "PandaPower metric snapshot",
            progress=progress,
            simulation_time=simulation_time,
            metadata=meta,
        )

    def _append_jsonl(self, name, event):
        if not self.results_dir:
            return
        event_dir = self.results_dir / "events"
        event_dir.mkdir(parents=True, exist_ok=True)
        with (event_dir / name).open("a", encoding="utf-8") as f:
            f.write(json.dumps(event, separators=(",", ":")) + "\n")
            f.flush()
            os.fsync(f.fileno())

    def _publish(self, event):
        if not self.producer:
            return
        try:
            self.producer.produce(self.EVENT_TOPIC, json.dumps(event), key=self.task_id)
        except Exception:
            pass

File: simservice/src/main/java/eu/fau/cs7/daceDS/SimService/SimulationEventEmitter.java
package eu.fau.cs7.daceDS.SimService;

import java.io.File;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.channels.FileLock;
import java.nio.charset.StandardCharsets;
import java.nio.file.StandardOpenOption;
import java.time.Instant;
import java.util.Map;
import java.util.UUID;

import org.apache.log4j.Logger;
import org.json.JSONObject;

import eu.fau.cs7.daceDS.Component.Config;
import eu.fau.cs7.daceDS.Kafka.ProducerImplKafka;

public class SimulationEventEmitter {
    private static final Logger logger = Logger.getLogger(SimulationEventEmitter.class.getName());
    private static final String EVENT_TOPIC = "simservice.logs.events";
    private static final Object FILE_WRITE_LOCK = new Object();
    private static ProducerImplKafka<String> producer;

    private final String taskId;
    private final String source;
    private final String component;

    public SimulationEventEmitter(String taskId, String source, String component) {
        this.taskId = taskId;
        this.source = source;
        this.component = component;
    }

    public void emit(String category, String eventCode, String message) {
        emit(category, eventCode, message, null, null, null);
    }

    public void emit(String category, String eventCode, String message, Double progress, Long simulationTime, Map<String, Object> metadata) {
        if (taskId == null || taskId.isEmpty()) {
            return;
        }

        JSONObject event = new JSONObject();
        event.put("id", "evt_" + UUID.randomUUID().toString());
        event.put("schema_version", "1.0");
        event.put("task_id", taskId);
        event.put("timestamp", Instant.now().toString());
        event.put("source", source);
        event.put("category", category);
        event.put("component", component);
        event.put("event_code", eventCode);
        event.put("message", message);
        if (progress != null) event.put("progress", progress);
        if (simulationTime != null) event.put("simulation_time", simulationTime);
        if (metadata != null) event.put("metadata", new JSONObject(metadata));

        String payload = event.toString();
        appendEvent(payload);
        publishEvent(payload);
    }

    private void appendEvent(String payload) {
        try {
            File eventDir = new File(getResultsDir(), taskId + "/events");
            eventDir.mkdirs();
            eventDir.setWritable(true, false);
            File eventFile = new File(eventDir, "structured_events.jsonl");
            File lockFile = new File(eventDir, ".writer.lock");
            synchronized (FILE_WRITE_LOCK) {
                try (FileChannel lockChannel = FileChannel.open(lockFile.toPath(),
                        StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
                    lockFile.setWritable(true, false);
                    try (FileLock lock = lockChannel.lock();
                         FileChannel writer = FileChannel.open(eventFile.toPath(),
                             StandardOpenOption.CREATE, StandardOpenOption.WRITE, StandardOpenOption.APPEND)) {
                        ByteBuffer bytes = StandardCharsets.UTF_8.encode(payload + "\n");
                        while (bytes.hasRemaining()) writer.write(bytes);
                        writer.force(true);
                    }
                }
            }
            eventFile.setWritable(true, false);
        } catch (Exception e) {
            logger.warn("Failed to append structured event for task " + taskId + ": " + e.getMessage());
        }
    }

    private void publishEvent(String payload) {
        try {
            ProducerImplKafka<String> p = getProducer();
            if (p != null) {
                p.publish(EVENT_TOPIC, taskId, payload, 0, 0, component);
            }
        } catch (Exception e) {
            logger.warn("Failed to publish structured event for task " + taskId + ": " + e.getMessage());
        }
    }

    private static synchronized ProducerImplKafka<String> getProducer() {
        if (producer == null) {
            producer = new ProducerImplKafka<String>("SimulationEventEmitter");
            producer.initPlainString();
        }
        return producer;
    }

    private String getResultsDir() {
        String resultsDir = Config.get("resultsDir");
        if (resultsDir == null || resultsDir.isEmpty()) {
            resultsDir = "/data/results";
        }
        return resultsDir;
    }
}
